### PR TITLE
Fix cppcheck 1.87 warnings in Geometry

### PR DESCRIPTION
**Pull request by soininen**
Fix most `cppcheck` 1.87 warnings in the Geometry module. Some obvious bugs got fixed as well.

**To test:**

Careful code review.

Fixes partially #22912.

*This does not require release notes* because this is an internal change.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by dtasev — DISMISSED** [review]
Looks good, removes some unused variable, and generally improves the code quality. Just the Windows build needs to pass

File: Framework/Geometry/src/Crystal/Group.cpp [hunk 1]
```diff
@@ -238,14 +237,11 @@ bool Group::hasIdentity() const {
...
 bool Group::eachElementHasInverse() const {
-  // Iterate through all operations, check that the inverse is in the group.
-  for (const auto &operation : m_allOperations) {
-    if (!containsOperation(getUnitCellIntervalOperation(operation.inverse()))) {
-      return false;
-    }
-  }
-
-  return true;
+  return std::all_of(m_allOperations.cbegin(), m_allOperations.cend(),
```
Comment (DanNixon) [inline Framework/Geometry/src/Crystal/Group.cpp]: :+1:

File: Framework/Geometry/src/Crystal/SymmetryElementFactory.cpp [hunk 2]
```diff
@@ -382,13 +382,15 @@ SymmetryElement_sptr SymmetryElementFactoryImpl::createFromPrototype(
...
     const SymmetryOperation &operation) const {
-  for (const auto &generator : m_generators) {
-    if (generator->canProcess(operation)) {
-      return generator;
-    }
+  const auto found = std::find_if(m_generators.cbegin(), m_generators.cend(),
+                                  [&operation](const auto &generator) {
+                                    return generator->canProcess(operation);
+                                  });
+  if (found == m_generators.end()) {
```
Comment (DanNixon) [inline Framework/Geometry/src/Crystal/SymmetryElementFactory.cpp]: Can this be changed to a ternary operator?

File: Framework/Geometry/src/MDGeometry/CompositeImplicitFunction.cpp [hunk 3]
```diff
@@ -47,10 +49,11 @@ std::string CompositeImplicitFunction::toXMLString() const {
   AutoPtr<Element> parameterListElement = pDoc->createElement("ParameterList");
   functionElement->appendChild(parameterListElement);
 
-  std::string functionXML;
-  for (const auto &Function : m_Functions) {
-    functionXML += Function->toXMLString();
-  }
+  const std::string functionXML =
+      std::accumulate(m_Functions.cbegin(), m_Functions.cend(), std::string(),
```
Comment (DanNixon) [inline Framework/Geometry/src/MDGeometry/CompositeImplicitFunction.cpp]: I think `std::accumulate` for string building was on the unclear side of the fence with respect to when STL algorithms should and should not be used.

**Review by DanNixon — CHANGES_REQUESTED** [review]
Just a couple of points.

File: Framework/Geometry/src/MDGeometry/CompositeImplicitFunction.cpp (hunk 3 above)
Comment (soininen) [reply Framework/Geometry/src/MDGeometry/CompositeImplicitFunction.cpp]: I missed that discussion. I will revert this change. I don't have a strong opinion on this anyway.

**Review by DanNixon — APPROVED** [review]

**Pull request merged**